### DISCUSSION TITLE
SDK Java lento / API lente ?

### DISCUSSION THREAD

**Discussion by demanzano** [fattureincloud/api#59] [Q&A]
Buonasera,

sto iniziando a fare la migrazione alle nuove API, utilizzo l'SDK Java ultima versione con JDK 11.

Tutto sommato i lavori procedono ma ho notato che le invocazioni alle web API sono MOLTO lente.

Per la creazione di nuovo cliente + creazione nuova proforma ad esso correlata ho atteso un totale di circa 15 secondi !! quindici !

Non so se sia lento il codice del SDK FiC  o proprio l'endpoint delle nuove API ma trovo che non abbia senso, le vecchie API sono quasi istantanee.

Mi sapete dire come mai ?

grazie

A. de Manzano

**Comment by valmoz** [comment]
Buongiorno,Ha provato ad effettuare la stessa chiamata dall'api reference? In questo modo potrà constatare l'effettiva velocità delle api.
GrazieMauro

Inviato da Yahoo Mail su Android 
 
  Il ven, 19 ago, 2022 alle 21:49, ***@***.***> ha scritto:   


Buonasera,

sto iniziando a fare la migrazione alle nuove API, utilizzo l'SDK Java ultima versione con JDK 11.

Tutto sommato i lavori procedono ma ho notato che le invocazioni alle web API sono MOLTO lente.

Per la creazione di nuovo cliente + creazione nuova proforma ad esso correlata ho atteso un totale di circa 15 secondi !! quindici !

Non so se sia lento il codice del SDK FiC o proprio l'endpoint delle nuove API ma trovo che non abbia senso, le vecchie API sono quasi istantanee.

Mi sapete dire come mai ?

grazie

A. de Manzano

**Comment by demanzano** [comment]
Buongiorno,

ho provato la create client dall'api reference, è pressochè istantanea come è giusto che sia, quindi come sospettavo è proprio l'SDK Java ad essere troppo lento.

Ho fatto delle prove invocando la stessa API (create client) da SDK, come minimo ci ha messo 4 secondi e un massimo di 8.
Ovviamente fatte varie prove, quindi escluso tempo di class loading, init dell'SDK,  eventuali interventi del JIT compiler ecc.

Non capisco come mai, dovrebbe essere solo un wrapper ad una chiamata http, cosa fa di cosi' lento ?

grazie

A. de Manzano

**Reply by emajo** [reply]
Buongiorno @demanzano, ho fatto un veloce test di creazione cliente con l'ultima versione della Java SDK e il tempo di risposta mi risulta inferiore ai 100 millisecondi, a questo punto credo che il problema risieda o nel suo codice o nella sua installazione di java.

Mi faccia sapere se riesce a trovare e risolvere il problema,
grazie,
Emanuele

**Reply by demanzano** [reply]
Buongiorno,

in effetti la cosa è davvero molto strana.
Il mio contesto di esecuzione è una applicazione Java EE 8 (con JDK 11) che gira dentro l'app server WildFly 21.x.
Il problema capita si in sviluppo locale che nel contesto di stage e di produzione.
Indago e provo a profilare le chiamate per cercare di capire dove si "ferma".
Non vorrei fosse qualche dipendenza-di-dipendenza (magari di Okhttp che vedo voi usate nel SDK) che fa qualcosa di "strano".

Per il momento in produzione son tornato alle vecchie API, iniziavo ad avere errori di timeout http in creazione fattura...

vi tengo informati.

grazie

A. de Manzano

**Reply by demanzano** [reply]
Buongiorno,

ho fatto delle prove invocando le API direttamente (via richieste HTTP, SENZA usare l'SDK di FiC), ma i risultati sono quasi identici.

Per la modifica di un cliente ci vogliono dai 2 ai 5 secondi, sempre e comunque.

Ho fatto una prova perfino con curl, bypassando Java ecc.  risultato identico.

Ho fatto una prova da una macchina in cloud (VM in Google Cloud) oltre che dalla mia locale (connessa FTTH a Milano, openfiber Aruba), risultato identico.

forse c'è qualcosa a livello di vs API gateway o auth server ? io uso l'autenticazione "manuale" con singolo (e fisso) Bearer token.

A disposizione se vi servono prove.

grazie

A. de Manzano

**Reply by emajo** [reply]
Ciao @demanzano questo potrebbe essere un problema legato al suo account, le chiedo gentilmente di contattare il nostro supporto clienti dalla chat presente sul nostro sito per poter andare così a controllare i log specifici del suo profilo.

Grazie,
Emanuele

**Reply by demanzano** [reply]
Buongiorno,

ok ho scritto alla live chat, vediamo cosa mi diranno, ho esposto il problema e fatto il tuo nome :)

grazie

A. de Manzano

**Reply by demanzano** [reply]
Buongiorno,

mi hanno risposto.. 

========
Deve sapere che la lentezza di tali azioni dipendono dal numero di documenti presenti nel suo account, come somma dei documenti gestiti nel corso degli anni di utilizzo del programma.
Purtroppo è normale che stia riscontrando dei rallentamenti a causa dei molti dati da caricare

I nostri tecnici sono a conoscenza di tale problematica e stanno lavorando per risolverla, è però purtroppo necessario pazientare
========

considerando che con le API v1 i tempi sono velocissimi mi sa che c'è qualcosa di "sbagliato" con le v2...
Spero che voi risolviate ASAP questo problema, sono sicuro di non essere l'unico ad avere tante fattture nell'account ;)

A. de Manzano
